### PR TITLE
!str #19129 New homes for file and java.io stream factories

### PR DESCRIPTION
**Pull request by johanandren**
Refs #19129

### CODE REVIEW

File: akka-docs-dev/rst/scala/migration-guide-1.0-2.x-scala.rst [hunk 1]
```diff
@@ -552,6 +556,12 @@ Example
       // This no longer works!
       val someFileSink = SynchronousFileSink(new File("."))
 
+      // This no longer works!
+      val yetAnotherFileSource = Source.file(new File("."))
```
Comment (drewhk) [inline akka-docs-dev/rst/scala/migration-guide-1.0-2.x-scala.rst]: remove these, the migration guide is from 1.0 to 2.0

**Comment by drewhk** [comment]
LGTM

**Comment by johanandren** [comment]
Found some migration docs I missed to update (on top of @drewhk s comment), will do tomorrow evening.

File: akka-docs-dev/rst/scala/migration-guide-1.0-2.x-scala.rst (hunk 1 above)
Comment (2m) [inline akka-docs-dev/rst/scala/migration-guide-1.0-2.x-scala.rst]: I think we can have it here, with a comment that it was like this in milestone versions. Might help users to migrate which are using milestone versions.

**Comment by 2m** [comment]
LGTM. I see that Dsl Consistency Spec is missing for these factories. I will create a PR to this PR to add that.

**Comment by drewhk** [comment]
There is the java migration guide too

**Comment by johanandren** [comment]
That's how I roll...

File: akka-bench-jmh-dev/src/main/scala/akka/stream/io/FileSourcesBenchmark.scala [hunk 2]
```diff
@@ -36,7 +36,7 @@ class FileSourcesBenchmark {
 
     val ft = Source(() ⇒ Iterator.continually(line))
       .take(10 * 39062) // adjust as needed
-      .runWith(Sink.file(f))
+      .runWith(FileIO.toFile(f))
```
Comment (ktoso) [inline akka-bench-jmh-dev/src/main/scala/akka/stream/io/FileSourcesBenchmark.scala]: For the record, I can live with those :wink:

File: akka-docs-dev/rst/java/code/docs/MigrationsJava.java [hunk 3]
```diff
@@ -202,16 +202,16 @@ public OutputStream create() {
...
-          Source.outputStream();
+          StreamConverters.asOutputStream();
 
         final Source<ByteString, OutputStream> otherOutputStreamSrc =
-          Source.outputStream(timeout);
+          StreamConverters.asOutputStream(timeout);
 
         final Sink<ByteString, InputStream> someInputStreamSink =
-          Sink.inputStream();
+          StreamConverters.asInputStream();
```
Comment (ktoso) [inline akka-docs-dev/rst/java/code/docs/MigrationsJava.java]: And those are very nice :+1:

File: akka-docs-dev/rst/java/migration-guide-1.0-2.x-java.rst [hunk 4]
```diff
@@ -544,14 +544,14 @@ should be replaced by
 InputStreamSource and OutputStreamSink
 ======================================
 
-Both have been replaced by ``Source.inputStream(…)`` and ``Sink.outputStream(…)`` due to discoverability issues.
+Both have been replaced by ``StreamConverters.fromInputStream(…)`` and ``Sink.fromOutputStream(…)`` due to discoverability issues.
```
Comment (rkuhn) [inline akka-docs-dev/rst/java/migration-guide-1.0-2.x-java.rst]: not `StreamConverters.fromOutputStream`?

File: akka-stream/src/main/scala/akka/stream/javadsl/FileIO.scala [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+import java.io.{ InputStream, OutputStream, File }
+
+import akka.japi.function
+import akka.stream.{ scaladsl, javadsl, ActorAttributes }
+import akka.util.ByteString
+
+import scala.concurrent.Future
+
+/**
+ * Java API: Factories to create sinks and sources from files
```
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/javadsl/FileIO.scala]: please remove `Java API`: separate APIs make this superfluous

**Comment by rkuhn** [comment]
PR validation failure is likely related, otherwise LGTM, thanks!

**Comment by ktoso** [comment]
LGTM, needs a rebase though

**Comment by johanandren** [comment]
Woho! Drinks for everyone!

**Comment by ktoso** [comment]
LGTM, :v:

**Pull request merged**